sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/aio/_client.py

```diff
@@ -0,0 +1,1179 @@
+                if attempt > 0:
+                    delay_seconds = self._reconnect_retry_policy.next_retry_delay(attempt)
+                    if not delay_seconds:
+                        break
+                    _LOGGER.debug(
+                        "Delay time for reconnect attempt %d: %ds",
+                        attempt,
+                        delay_seconds,
+                    )
+                    await asyncio.sleep(delay_seconds)
+                await self._start_from_restarting(attempt)
+                if self._ws:
+                    self._ws.clear_reconnect_recover()
+                success = True
+                break
+            except ReconnectError:
+                _LOGGER.debug("fail to reconnect, and will retry in another task")
+                break
+            except Exception as e:  # pylint: disable=broad-except
+                _LOGGER.warning("An attempt to reconnect connection failed: %s", e)
+        if not success:
+            await self._handle_connection_stopped()
+        else:
+            _LOGGER.debug("reconnect successfully")
+
+    async def _handle_connection_stopped(self):
+        _LOGGER.debug("Connection stopped")
+        self._is_stopping = False
+        self._state = WebPubSubClientState.STOPPED
+        self._call_back(CallbackType.STOPPED)
+
+    async def _handle_connection_close_and_no_recovery(self, reconnect_tried_times: Optional[int] = None):
+        _LOGGER.debug("Connection closed and no recovery")
+        self._state = WebPubSubClientState.DISCONNECTED
+        self._call_back(
+            CallbackType.DISCONNECTED,
+            OnDisconnectedArgs(
+                connection_id=self._connection_id,
+                message=(self._last_disconnected_message.message if self._last_disconnected_message else None),
+            ),
+        )
+        if self._auto_reconnect:
+            await self._handle_auto_reconnect(reconnect_tried_times)
+        else:
+            await self._handle_connection_stopped()
+
+    def _is_connected(self) -> bool:
+        """check whether the client is still connected to server after open
+
+        :return: True if the client is connected to server, otherwise False
+        :rtype: bool
+        """
+        return self._state == WebPubSubClientState.CONNECTED and self._ws and not self._ws.closed  # type: ignore
+
+    async def _rejoin_group(self, group_name: str):
+        async def _rejoin_group_func():
+            try:
+                await self._join_group_core(group_name)
+                _LOGGER.debug("rejoin group %s successfully", group_name)
+            except Exception as e:  # pylint: disable=broad-except
+                _LOGGER.debug("fail to rejoin group %s", group_name)
+                self._call_back(
+                    CallbackType.REJOIN_GROUP_FAILED,
+                    OnRejoinGroupFailedArgs(group=group_name, error=e),
+                )
+
+        await _rejoin_group_func()
```

Comment:
I'm confused why we're creating a nested function, then immediately calling that function?

Reply:
Kashif advise we need try to make code structure same between async and sync API. Maybe it is too odd here for async API.

Reply:
Updated